panel/auth.py

```diff
@@ -388,6 +418,126 @@ def _USER_KEY(self):
         return config.oauth_extra_params.get('USER_KEY', os.environ.get('PANEL_USER_KEY', 'email'))
 
 
+class PasswordLoginHandler(GenericLoginHandler):
+
+    _EXTRA_TOKEN_PARAMS = {
+        'grant_type': 'password'
+    }
+
+    def get(self):
+        try:
+            errormessage = self.get_argument("error")
+        except Exception:
+            errormessage = ""
+
+        next_url = self.get_argument('next', None)
+        if next_url:
+            self.set_cookie("next_url", next_url)
+        html = self._basic_login_template.render(
+            errormessage=errormessage,
+            PANEL_CDN=CDN_DIST
+        )
+        self.write(html)
+
+    async def post(self):
+        username = self.get_argument("username", "")
+        password = self.get_argument("password", "")
+        params = {
+            "scope": ' '.join(self._SCOPE),
+            "grant_type": "password",
+            "client_id": config.oauth_key,
+            "username": username,
+            "password": password,
+            **self._EXTRA_TOKEN_PARAMS
+        }
+
+        http = self.get_auth_http_client()
+
+        # Request the authorization token.
+        req = HTTPRequest(
+            self._OAUTH_ACCESS_TOKEN_URL,
+            method='POST',
+            body=urlparse.urlencode(params),
+            headers=self._API_BASE_HEADERS
+        )
+
+        try:
+            response = await http.fetch(req)
+        except HTTPError as e:
+            return self._on_error(e.response)
+
+        body = decode_response_body(response)
+
+        if 'access_token' not in body:
+            return self._on_error(response, body)
+
+        user_headers = dict(self._API_BASE_HEADERS)
+        if self._access_token_header:
+            user_url = self._OAUTH_USER_URL
+            user_headers['Authorization'] = self._access_token_header.format(
+                body['access_token']
+            )
+        else:
+            user_url = '{}{}'.format(self._OAUTH_USER_URL, body['access_token'])
+
+        user_response = await http.fetch(user_url, headers=user_headers)
+        user = decode_response_body(user_response)
```

Comment:
In our use case, The user url is a custom URL which  uses one of the decoded properties of the access token as one of its path parameters. We can't supply a fixed url.
We actually don't care about the user info so much, either Make this optional, or provide a way (like a hook) to set the user info ourselves

Reply:
I'll have to retest this but when I tried it with your standard endpoint I got back a valid response containing my user information.

Reply:
Did you set `openid` as one of the scopes?

Reply:
I couldn't reach that point due to the infinite redirects bug